arangod/Aql/ClusterQuery.cpp

```diff
@@ -45,43 +46,44 @@ using namespace arangodb::aql;
 ClusterQuery::ClusterQuery(QueryId id,
                            std::shared_ptr<transaction::Context> ctx,
                            QueryOptions options)
-    : Query(id, ctx, aql::QueryString(),
-            /*bindParams*/ nullptr, std::move(options),
+    : Query{id,
+            ctx,
+            {},
+            /*bindParams*/ nullptr,
+            std::move(options),
             /*sharedState*/ ServerState::instance()->isDBServer()
                 ? nullptr
-                : std::make_shared<SharedQueryState>(ctx->vocbase().server())) {
+                : std::make_shared<SharedQueryState>(ctx->vocbase().server())} {
 }
 
-ClusterQuery::~ClusterQuery() {
+void ClusterQuery::destroy() noexcept {
   try {
     _traversers.clear();
   } catch (...) {
   }
+  Query::destroy();
 }
 
 /// @brief factory method for creating a cluster query. this must be used to
 /// ensure that ClusterQuery objects are always created using shared_ptrs.
-/*static*/ std::shared_ptr<ClusterQuery> ClusterQuery::create(
+std::shared_ptr<ClusterQuery> ClusterQuery::create(
     QueryId id, std::shared_ptr<transaction::Context> ctx,
-    aql::QueryOptions options) {
-  // workaround to enable make_shared on a class with a private/protected
-  // constructor
-  struct MakeSharedQuery : public ClusterQuery {
-    MakeSharedQuery(QueryId id, std::shared_ptr<transaction::Context> ctx,
-                    aql::QueryOptions options)
-        : ClusterQuery(id, std::move(ctx), std::move(options)) {}
-  };
-
+    QueryOptions options) {
   TRI_ASSERT(ctx != nullptr);
-
-  return std::make_shared<MakeSharedQuery>(id, std::move(ctx),
-                                           std::move(options));
+  return std::shared_ptr<ClusterQuery>{
+      new ClusterQuery{id, std::move(ctx), std::move(options)},
+      [](ClusterQuery* query) noexcept {
+        if (query) {
+          query->destroy();  // to prevent data race on vptr
+          delete query;
+        }
+      }};
```

Comment:
That won't work - we have to use `make_shared` because otherwise the internal weak_ptr in `enabled_shared_from_this` won't be initialized correctly. (I tried something similar before and learned this the hard way.)

But that is actually not necessary. It should be sufficient if each _most derived_ subclass made sure that the query is removed from the QueryList (i.e., calls `QueryProfile::unregisterFromQueryList()`).

Reply:
Yes it should work.

Reply:
Is enabled_shared_from_this cause some initialize btw?
I think your suggestion is better. But I think old code should work if we cal shared/weak_from_this() after ctor